storage/innobase/handler/ha_innodb.cc

```diff
@@ -15741,6 +15741,22 @@ ha_innobase::extra(
 		break;
 	case HA_EXTRA_END_ALTER_COPY:
 		trx = check_trx_exists(ha_thd());
+		if (m_prebuilt->table->skip_alter_undo) {
+			if (dberr_t err= trx->bulk_insert_apply()) {
+				if (err == DB_DUPLICATE_KEY) {
+					return HA_ERR_FOUND_DUPP_KEY;
+				}
+				return 1;
+			}
+
+			trx->end_bulk_insert(*m_prebuilt->table);
+			trx->bulk_insert = false;
```

Comment:
How will this work on a partitioned table? Each partition would be a separate table from the InnoDB point of view.

Reply:
All the partition will do bulk insert operation.
```
--source include/have_innodb.inc
--source include/have_partition.inc
CREATE TABLE t1 ( i INT NOT NULL AUTO_INCREMENT PRIMARY KEY, f INT )
        ENGINE = InnoDB PARTITION BY HASH(i) PARTITIONS 2;
INSERT INTO t1 VALUES (2, 2), (3, 3), (4, 4), (5, 5);
ALTER TABLE t1 ADD PARTITION PARTITIONS 2;
DROP TABLE t1;
```
In the above alter, t1 will get rebuild. All the partition will do bulk insert operation. Each partition
will call `HA_EXTRA_BEGIN_ALTER_COPY` and `HA_EXTRA_BEGIN_ALTER_COPY`